implement: new configurable script-tag approach

chase-moskal opened the pull request:
hey @adamsilverstein 

i was playing around on your branch, with some experimental ideas to improve the developer-experience for installing webp-hero, and i was hoping you could review the results here

this commit

- removes the passing of any `window` globals
- removes any default hardcoded links
- removes any entanglement with polyfills
- uses this "data attributes as settings" approach that i find novel and fascinating
- only has an html installation pathway

the idea is to encourage new users of webp-hero to use conditionally via this simple html script tag installation.

please let me know what you think.

what follows is the commit message https://github.com/chase-moskal/webp-hero/commit/71bdfb907f9b9fd05cc9527759597f6d0ac4ef2e and finally some closing thoughts 

------

this will be the simplest way for any technician to install webp-hero.

a script tag with some attributes is all that's required:

```html
<script
  defer
  data-webp-hero
  src="https://unpkg.com/webp-hero@0.0.3/dist-cjs/conditionally.js"
></script>
```

by default, the url to the webp-hero bundle is "contrived" by reading
this script element's `src` attribute, using a little regex.

the technician may provide additional attributes,
as optional settings:

 - `data-bundle`  
   explicity set the url to load the webp-hero bundle.

 - `data-use-custom-behavior`  
   instruct "conditionally" to "do nothing" after loading the bundle.
   this allows a developer to use webp hero bundle in a custom way.
   if not provided, `polyfillDocument` will be called.

 - `data-force` *(for testing)*  
   force the webp-hero bundle to be loaded, by lying to webp-hero and
   reporting that webp is not supported (even if it is).

 - `data-force-use-canvas-elements` *(for testing)*  
   force the usage of canvas elements polyfill technique.
   only applies when `data-use-custom-behavior` is *NOT* set.

additionally, there's a `webpHeroPromise`, for implementing custom
behaviors:

 - the `conditionally` script always sets `window.webpHeroPromise`.

 - if webp-hero is loaded,  
   the promise resolves with the `webpHero` module object.

 - otherwise if webp-hero was not loaded,  
   the promise resolves with `undefined`.

------

please give this a review, let me know what you think of this approach, and let me know if these changes would satisfy your usage requirements.

if you approve, i will take the next step of writing changes to the readme and related housekeeping, and will publish in `v0.0.3`.

chase-moskal commented:
oh, and if we decide to move forward with this, of course, i'll test it in many browsers

adamsilverstein commented on index.html:
```diff
@@ -36,29 +36,37 @@ <h1>webp-hero</h1>
 		<img alt="" src="data:image/webp;base64,UklGRhgFAABXRUJQVlA4WAoAAAAgAAAANgAAJwAASUNDUBgCAAAAAAIYAAAAAAQwAABtbnRyUkdCIFhZWiAAAAAAAAAAAAAAAABhY3NwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAA9tYAAQAAAADTLQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAlkZXNjAAAA8AAAAHRyWFlaAAABZAAAABRnWFlaAAABeAAAABRiWFlaAAABjAAAABRyVFJDAAABoAAAAChnVFJDAAABoAAAAChiVFJDAAABoAAAACh3dHB0AAAByAAAABRjc ...
 
 		<script src="dist-cjs/polyfills.js"></script>
```
Is `polyfills.js` required? can it also be conditionally loaded or maybe built into `webp-hero.bundle.js` directly?

Ideally we would only need to enqueue `conditionally.js` to get it to work, and if the browser already supported WebP nothing else would load.

chase-moskal replied:
the polyfills are required by IE11 and older, and surely other old browsers too.

you raise a good point!

i think a good solution might be a simple new attribute, `data-include-polyfills`, which would cause *conditionally* to load the polyfills prior to the webp-hero bundle.

adamsilverstein replied:
👍🏼 that sounds like a good option. It would also be helpful to document exactly when the polyfills are needed, maybe a list of specific browsers that require it? I could help with a follow up task and log that info while doing browserstack testing

chase-moskal replied:
excellent idea, i'll try to generate a list of which browsers benefit from the polyfills for the readme :+1:

adamsilverstein commented on source/conditionally.ts:
```diff
@@ -1,24 +1,41 @@
+
+window.webpHeroPromise = Promise.resolve().then(async() => {
+	const selector = "script[data-webp-hero]"
+	const settings = getScriptSettingsForConditionally({selector})
```
Nice!

adamsilverstein commented on source/utils/get-script-settings-for-conditionally.ts:
```diff
@@ -0,0 +1,47 @@
+}
+
+function contriveBundleUrlFromConditionallySrc(src: string) {
+	const [,baseUrl] = src.match(/^(.*)\/conditionally\.js$/) ?? []
```
👍🏼

adamsilverstein commented:
overall love it, well done! I left one question around the polyfill file here - https://github.com/adamsilverstein/webp-hero/pull/1/files#r882276447 otherwise good to go after some testing!

adamsilverstein commented:
> oh, and if we decide to move forward with this, of course, i'll test it in many browsers

👍🏼 Happy to test a variety of browser as well using Browserstack.

chase-moskal commented:
awesome, thanks for such a prompt review @adamsilverstein!

i should have some time this week, or this weekend, to address the polyfill situation by implementing `data-include-polyfills`, and then doing the necessary work to release all of this into a fresh version of webp-hero! 🦸‍♂️ 

thanks you for your thoughtful contributions, comments, and reviews throughout this process -- i'll ping you when more progress is made here :+1:

chase-moskal commented:
hey @adamsilverstein 

i made some more progress on this. i've implemented the optional attribute `data-include-polyfills`, and did some refactoring.

i've also reworked the readme, featuring the all-new much-simpler installation instructions — you can [read it here](https://github.com/chase-moskal/webp-hero/tree/conditionally#readme).
- of course, the install snippet shown there doesn't actually work yet (v0.1 isn't published yet!)

unless you have any reservations about these changes or the new readme, i'll merge this into master soon. we need to keep a keen eye out for any mistakes that could cause any breaking changes for existing users (i hope i haven't missed anything).

afterwards, i think the next steps will be:
1. update all the npm dependencies (like typescript and such)
1. update/rebuild emscripten and libwebp
1. update the `CHANGELOG.md`
1. publish webp-hero `v0.1.0-rc.1`
1. test against every available browser (i'm also using browserstack)
1. verify that `rc.1` works, then publish `v0.1.0`

:beers: your feedback and review is much appreciated

adamsilverstein commented:
> i made some more progress on this. i've implemented the optional attribute data-include-polyfills, and did some refactoring.

Thanks for the update @chase-moskal - I am back and will review your most recent changes, thanks for working on this!

chase-moskal commented:
hey @adamsilverstein — i was thinking of proceeding to test and publish this work soon — just thought i'd ping you to see if you have feedback beforehand, i wouldn't want to miss it!

:beers:

adamsilverstein commented:
Thanks for the ping, I'll give a review this week.
